### ISSUE TITLE
Incompatibility between Spring Cloud Stream Binder Rabbit and spring-amqp after upgrading Spring Boot dependencies to 3.4.0

### ISSUE THREAD

**Issue by polyoxidonium** [spring-cloud/spring-cloud-stream#3049]
### **Issue Overview:**
After upgrading Spring Boot dependencies BOM to version 3.4.0 and Spring Cloud dependencies to 2023.0.3, we encountered an incompatibility issue where Spring Cloud Stream Binder Rabbit (version 4.1.3) pulled in spring-amqp version 3.2.x. This caused a method resolution error due to the following method being removed in spring-amqp 3.2.x:
`org.springframework.amqp.core.ExchangeBuilder.durable(boolean)`

The error occurs when RabbitExchangeQueueProvisioner attempts to use the durable(boolean) method, which was moved from a child class to an abstract class in spring-amqp 3.2.x.

The issue appears due to spring-amqp version 3.2.x being included from the Spring Boot BOM, which is incompatible with the method usage in Spring Cloud Stream Binder Rabbit.

### **Steps to Reproduce:**

1. Upgrade Spring Boot dependencies BOM to 3.4.0 (which includes spring-amqp:3.2.0).
2. Upgrade Spring Cloud dependencies BOM to 2023.0.3.
3. Use spring-cloud-stream-binder-rabbit (version 4.1.3 or later) without explicitly defining the AMQP binder.
4. Run the application, and the following error occurs:
```
An attempt was made to call a method that does not exist. The attempt was made from the following location:
    org.springframework.cloud.stream.binder.rabbit.provisioning.RabbitExchangeQueueProvisioner.buildExchange(RabbitExchangeQueueProvisioner.java:654)

The following method did not exist:
    'org.springframework.amqp.core.ExchangeBuilder org.springframework.amqp.core.ExchangeBuilder.durable(boolean)'
```

### **Expected Behavior:**
The application should function correctly with no method resolution errors when using spring-amqp 3.2.x. If there are breaking changes in the spring-amqp API, they should be properly handled, and the correct version should be included to avoid conflicts.

### **Workaround:**
We were able to resolve the issue by explicitly downgrading spring-amqp to version 3.1.8. However, it would be ideal to handle this conflict in a way that ensures compatibility across versions.

### **To fix this:**
**Potential solution 1:**  Modify spring-cloud-stream-binder-rabbit to properly align with spring-amqp versions and avoid pulling in incompatible versions of spring-amqp.
**Potential solution 2:**  Ensure that method calls in RabbitExchangeQueueProvisioner are updated to reflect changes made in spring-amqp 3.2.x or earlier versions.

### **Versions Used:**
Spring Boot BOM: 3.4.0 (which includes spring-amqp:3.2.0)
Spring Cloud Stream Binder Rabbit: 4.1.3
Spring AMQP: 3.2.0 (problematic version)
Spring Cloud: 2023.0.3

**Comment by polyoxidonium** [comment]
> I believe this is due to bytecode incompatibility: the method is there, but it is in the super class. So, to make it working , that old Spring Cloud version has to be compiled against latest Spring AMQP. That’s why we didn’t notice any difference from current main: no need in code change - works because we have compiled it against Spring AMQP 3.2.
> 
> We will think next week what we can do. Meantime I’d like to know if there was really a claim that Spring Cloud `2023` is supposed to be compatible with just released Spring Boot `3.4`. Maybe we all just need to wait when Spring Cloud `2024` is released: https://github.com/spring-cloud/spring-cloud-release/milestone/147?

Thanks for the reply. 
You are right, official claim is that 2023 is supposed to be compatible with 3.3.x and 3.2.x

As you stated, recompilation is necessary so I hope 2024 will already make the job so we can use latest BOMs from both.

Thank you again for the feedback.

**Comment by artembilan** [comment]
Well, I definitely can treat this as a duplicate of: https://github.com/spring-cloud/spring-cloud-stream/issues/3048.
Since looks like everyone tries to upgrade to the latest Spring Boot, while staying on the previous Spring Cloud.
As we have agreed this is not what has been claimed, so better to upgrade to the latest Spring Cloud when it is available.
